xapian-applications/omega/query.cc

```diff
@@ -651,24 +652,28 @@ html_strip(const string &str)
     return res;
 }
 
-// FIXME split list into hash or map and use that rather than linear lookup?
-static int word_in_list(const string& word, const string& list)
-{
-    string::size_type split = 0, split2;
+static unordered_map<string, int> word_to_occurrence;
+
+static void build_word_map(const string& list) {
+    string word;
     int count = 0;
-    while ((split2 = list.find('\t', split)) != string::npos) {
-	if (word.size() == split2 - split) {
-	    if (memcmp(word.data(), list.data() + split, word.size()) == 0)
-		return count;
-	}
-	split = split2 + 1;
-	++count;
+    for (size_t i = 0; i <= list.length(); ++i) {
+	if (list[i] == '\t' || list[i] == '\0') {
+	    if (word_to_occurrence.find(word) == word_to_occurrence.end()) {
+		word_to_occurrence[word] = count++;
+		word.clear();
+	    }
+	} else
+	    word += list[i];
     }
-    if (word.size() == list.size() - split) {
-	if (memcmp(word.data(), list.data() + split, word.size()) == 0)
-	    return count;
-    }
-    return -1;
+}
+
+static int word_in_list(const string& word) {
+    int pos = -1;
+    if (word_to_occurrence.find(word) != word_to_occurrence.end())
+	pos = word_to_occurrence[word];
+    word_to_occurrence.clear();
```

Comment:
Doesn't this guarantee that the second call to `word_in_list()` below will always return `-1`? That suggests to me that some more tests are needed.

Additionally, it might be clearer to do this by returning the word list `unordered_map` from `build_word_map()`, and explicitly freeing it in `html_highlight()` (or, perhaps better, heap allocate it within `html_highlight()` and pass it in to the other functions).